### PR TITLE
Add block size for paged blocks

### PR DESCRIPTION
**Pull request by dbw9580**
### What changes are proposed in this pull request?

Put block size metadata in paged blocks, so that the information can persist across worker restarts.

### Why are the changes needed?

The block size is vital for various worker operations. However, page store does not currently support attaching metadata to pages, therefore this metadata is lost when the worker restarts and all pages are invalidated. It would be beneficial that the pages are preserved.

### Does this PR introduce any user facing changes?

No.

### CODE REVIEW

File: core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java
Comment (jja725) [inline core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java:100]: in which situation we would expect empty? I feel like if there's no compatibility concern it should be runtime exception which indicate there must be something wrong if we can't parse the blockId

File: core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java
Comment (jja725) [inline core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java:157]: It's a bit weird here since NumberFormatException is extending IllegalArgumentException. And I'm not a big fan of adding RuntimeException in method signature, java doc should be enough.

Comment (jja725) [reply core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java:157]: And I'm also a little confused about expectation here. tryDowncast means we are expecting failure, which seems like Optional is a better option. But doesn't seems like we explicitly catch exception and handle it. So maybe we should just name it as downcast.

**Review by jja725 — COMMENTED** [review]
Overall LGTM, leave some minor questions about exceptions.

File: core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java
Comment (dbw9580) [reply core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java:165]: > tryDowncast means we are expecting failure, which seems like Optional is a better option.

Yeah the intention was to signal to the caller that downcasting failure is expected, rather than an "exception". Using an `Optional`, however, does not give you a detailed error message of what went wrong, what's the part that does not match a paged block ID.

But now since I came to rethink about it, all this code is in the worker's page store, so there should not be any page that does not follow the name pattern. Anything that does is really an exception.

Comment (dbw9580) [reply core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java]: fixed in [3206ba3](https://github.com/Alluxio/alluxio/pull/16254/commits/3206ba3bb61e5dfee62a9f244827552eb7e51d70)

File: core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java
Comment (jja725) [inline core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java:167]: It's a bit weird that we catch and throw the same exception here... better we just bypass the IllegalArgumentException from parseBlockId/parseBlockSize

File: core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java
Comment (beinan) [inline core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java:29]: I would use a simpler template %016x_%016x to reduce the parsing cost, your call.

**Review by beinan — APPROVED** [review]
lgtm

File: core/server/worker/src/main/java/alluxio/worker/grpc/BlockReadHandler.java
Comment (dbw9580) [inline core/server/worker/src/main/java/alluxio/worker/grpc/BlockReadHandler.java:527]: this needs rebasing after #16322 is merged

**Comment by beinan** [comment]
#16322 got merged,  rebase?  @dbw9580

**Review by jja725 — APPROVED** [review]

**Comment by dbw9580** [comment]
@beinan @jja725 rebased on latest change, PTAL thanks!

File: core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java
Comment (dbw9580) [reply core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java:29]: My concern is that if the user somehow misconfigures client local cache and worker page store to share the same cache directory, the special naming scheme is the only thing that may help prevent we messing up the pages, as other apps may pretty well also use a naming scheme as simple as `number_number`.

**Review by beinan — APPROVED** [review]

**Comment by beinan** [comment]
alluxio-bot, merge this please.

**Pull request merged**
